---
## Hausaufgabe 3 - Verbesserungsverfahren
---

### Achtung: Bitte überprüfen Sie vor Abgabe der Hausaufgabe, ob das Notebook richtig gespeichert wurde. 
Die Speicherung von Notebooks funktioniert die Menüleiste oder über die Shortcuts Strg + S (Win) und CMD + S (MAC).

---
## Bewertung
---

#### Erreichbare Punkte: 25
---
#### Erreichte Punkte:   -
---
---
Aufgabe 1:<br>
<br>
Aufgabe 2:<br>
<br>
Aufgabe 3:<br>

#### Aufgabe 1 - Theorie I: TwoEdgeExchange Move (2 Punkte)

Der 2-opt Algorithmus ist ein sehr bekanntes Verbesserungverfahren aus der Tourenplanung. Im Kern werden jeweils zwei **nicht-benachbarte** Kanten entfernt und die so entstandenen Teilsequenzen neu verbunden. Ein Beispiel ist in dem folgenden Bild dargestellt:

<img src="2-opt.png" alt="Drawing" style="width: 300px;"/>

Dieses Konzept kann auch auf das Scheduling angewandt werden. Dabei werden in einer Auftragsreihenfolge nicht nur die Aufträge als einzelne Elemente betrachtet, sondern auch deren Verbindungen. Wenn Auftrag $i$ direkt auf Auftrag $j$ folgt, so sind diese beiden Aufträge über eine Verbindungen, bzw. "Kante", $(i,j)$ verbunden. Somit lässt sich der sog. TwoEdgeExchange Move wie folgt definieren:

Ein TwoEdgeExchange Move $M(\text{2-EdgeExchange}, s, i, j)$ beschreibt den Austausch der Verbindungen zweier Auftragspaare $(i,i+1)$ und $(j,j+1)$. Dabei dürfen die Verbindungen **nicht-benachbart** sein und es werden **nur** die Verbindungen $(i,i+1)$ und $(j,j+1)$ entfernt und die Aufträge neu zu $(i,j)$ und $(i+1,j+1)$ zusammengefügt. 

Es ist folgende Auftragsreihenfolge gegeben `s=[0,1,2,3,4,5,6]`. Führen Sie (händisch) folgenden TwoEdgeExchange-Move durch $M(\text{2-EdgeExchange}, s, 1, 4)$.

**Hinweis:** Das erste Element steht an der Position 0.

$M(\text{2-EdgeExchange}, s, 1, 4) = ??$ 
[0, 1, 4, 3, 2, 5, 6]

#### Aufgabe 2 - Theorie II: TwoEdgeExchange Move (5 Punkte)

**a)** In der Abbildung in Aufgabe 1 werden die Verbindungen zwischen den Auftragspaaren (0,1) und (2,3) ausgewählt. Nach den Autoren Kirlik und Oguz lässt sich jedoch der TwoEdgeExchange Move weiter einschränken, indem sich die Positionen $i$ und $j$ um mindestens 3 Position $\left(|j-i|\geq 3\right)$ unterscheiden müssen. Was könnte der Grund für diese weitere Einschränkung sein. 

**Hinweis:** Beachten Sie dabei die bereits umgesetzten Nachbarschaften.

**Literaturverweis** (in OPAL vorhanden): Kirlik, G., & Oguz, C. (2012). A variable neighborhood search for minimizing total weighted tardiness with sequence dependent setup times on a single machine. Computers & Operations Research, 39(7), 1506-1520.

Dadurch wird zum einen die Laufzeit des Rechners reduziert, um die Effizienz der Erzielung einer lokal optimalen Lösung zu verbessern. Andererseits teilt der Ansatz der Autoren die ursprünglichen (n-1) Kanten in zwei Gruppen auf, wobei die erste Gruppe die ersten drei Kanten und die andere Gruppe die hinter liegenden (n-1-3) Kanten umfasst. Dadurch wird sichergestellt, dass beide Kanten in der Solution sowohl vorwärts als auch rückwärts angeordnet sind. Dies reicht aus, um die lokal optimale Lösung zu finden, da wir bei der Suche nach der lokal optimalen Lösung das Gleichgewicht zwischen der Rechenzeit und der Perfektion des Ergebnisses berücksichtigen müssen.


**b)** Bestimmen Sie die Komplexität der TwoEdgeExchange-Move Nachbarschaft. Nehmen Sie an, dass eine Lösung durch eine Permutation der Länge $n$ repräsentiert ist und $|j-i|\geq 3$ gilt.

**Hinweise:** Geben Sie die Komplexität in Abhängigkeit von $n$ an.

**Tipp:**
<details>

Wie viele Kanten sind in einer Sequenz enthalten? Wie viele Kanten können getauscht werden, dass eine neue Lösung erzielt wird? Wieviele Dopplungen/Symmetrien gibt es?
</details>

$ |N(\text{2-EdgeExchange}, s)| = ??$
1/2 * (n-4)*(n-3)

#### Aufgabe 3 - Implementierung: TwoEdgeExchange-Move (18 Punkte)

Erweitern Sie den in der Lehrveranstaltung erstellten Solver um eine TwoEdgeExchange-Move Nachbarschaft. Erstellen Sie die Klassen `TwoEdgeExchangeMove` und `TwoEdgeExchangeNeighborhood`.



**Hinweise:** 

Implementieren Sie ihre Lösung in den bereitgestellten `.py` Dateien aus der Veranstaltung, nicht hier im Notebook. Bitte kopieren Sie der Übersichtlichkeit wegen trotzdem Ihre in der `.py`-Datei implementierten TwoEdgeExchange-Move Komponenten hier ins Notebook. Geben Sie alle zum fehlerfreien Ausführen notwendigen Dateien an.


Testen Sie die Nachbarschaft auf dem Datensatz `InputFlowshopSIST20.json`. Verwenden Sie die BestImprovement-Strategie und die FCFS-Startlösung. Mögliches erwartetes Ergebnis:


In [33]:
list1 = [1,2,3]
list2 = [11,22,33]
list3 = [111,222,333]

# print(list1.extend(list2).extend(list3))

list1.extend(list2)
list1.extend(list3)
print(list1)
print(list1[:3])
print(list1[0:3])

[1, 2, 3, 11, 22, 33, 111, 222, 333]
[1, 2, 3]
[1, 2, 3]


In [1]:
from Solver import *


data = InputData("InputFlowshopSIST20.json")

localSearch = IterativeImprovement(data, 'BestImprovement', ['TwoEdgeExchange'])

solver = Solver(data, 1008)

solver.RunLocalSearch(constructiveSolutionMethod='FCFS', algorithm=localSearch)

Generating an initial solution according to FCFS.
Constructive solution found.
The permutation [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1482


TypeError: 'NoneType' object is not iterable

          Generating an initial solution according to FCFS.
          Constructive solution found.
          The permutation [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1482
          The permutation [0, 1, 10, 9, 8, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1375
          The permutation [8, 9, 10, 1, 0, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1360
          The permutation [8, 9, 10, 1, 0, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 16, 15, 17, 18, 19] results in a Makespan of 1341
          The permutation [9, 8, 10, 1, 0, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 16, 15, 17, 18, 19] results in a Makespan of 1340
          The permutation [9, 8, 10, 1, 0, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 16, 17, 15, 18, 19] results in a Makespan of 1334
          The permutation [9, 8, 10, 1, 0, 7, 6, 5, 4, 12, 11, 2, 3, 13, 14, 16, 17, 15, 18, 19] results in a Makespan of 1312
          The permutation [9, 8, 10, 1, 0, 7, 6, 5, 4, 12, 17, 16, 14, 13, 3, 2, 11, 15, 18, 19] results in a Makespan of 1265
          The permutation [9, 1, 10, 8, 0, 7, 6, 5, 4, 12, 17, 16, 14, 13, 3, 2, 11, 15, 18, 19] results in a Makespan of 1256
          Reached local optimum of TwoEdgeExchange neighborhood. Stop local search.
          Best found Solution.
          The permutation [9, 1, 10, 8, 0, 7, 6, 5, 4, 12, 17, 16, 14, 13, 3, 2, 11, 15, 18, 19] results in a Makespan of 1256